### PR TITLE
mam/mss: sign and advance skn

### PR DESCRIPTION
**Pull request by tsvisabo**

### CODE REVIEW

File: common/errors.h [hunk 1]
```diff
@@ -475,6 +475,7 @@ enum retcode_t {
   RC_MAM_SEND_CTX_NOT_FOUND = 0x19 | RC_MODULE_MAM | RC_SEVERITY_MODERATE,
   RC_MAM_MESSAGE_NOT_FOUND = 0x1A | RC_MODULE_MAM | RC_SEVERITY_MODERATE,
   RC_MAM_BAD_PACKET_ORD = 0x1B | RC_MODULE_MAM | RC_SEVERITY_MODERATE,
+  RC_MAM_MSS_EXAHUSTED = 0x1C | RC_MODULE_MAM | RC_SEVERITY_MODERATE,
```
Comment (thibault-martinez) [inline common/errors.h]: typo

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

File: mam/mss/mss.c [hunk 2]
```diff
@@ -537,6 +550,14 @@ bool mam_mss_next(mam_mss_t *mss) {
   return true;
 }
 
+size_t mam_mss_num_remaining_sks(mam_mss_t *mss) {
+  if (mss->skn >= MAM_MSS_MAX_SKN(mss->height)) {
+    return 0;
+  }
+
+  return mss->skn - MAM_MSS_MAX_SKN(mss->height);
```
Comment (thibault-martinez) [inline mam/mss/mss.c]: Should be the opposite ?

File: mam/mss/mss.c [hunk 3]
```diff
@@ -537,6 +550,14 @@ bool mam_mss_next(mam_mss_t *mss) {
   return true;
 }
 
+size_t mam_mss_num_remaining_sks(mam_mss_t *mss) {
```
Comment (thibault-martinez) [inline mam/mss/mss.c]: constness

File: mam/mss/mss.h [hunk 4]
```diff
@@ -192,12 +192,39 @@ void mam_mss_auth_path(mam_mss_t *mss, mss_mt_idx_t skn, trits_t path);
...
+ * Advances skn
+ *
+ * @param mss [in] MSS interface
+ *
+ * @return True if can produce next signature
+ */
 bool mam_mss_next(mam_mss_t *mss);
 
+/**
+ * Advances skn
```
Comment (thibault-martinez) [inline mam/mss/mss.h]: Wrong def

File: mam/api/api.c [hunk 5]
```diff
@@ -288,6 +292,26 @@ retcode_t mam_api_bundle_write_packet(
   return RC_OK;
 }
 
+retcode_t mam_api_num_remaining_sks(mam_api_t *const api,
```
Comment (thibault-martinez) [inline mam/api/api.c]: Should this be done on a channel/endpoint instead ?

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

File: mam/mss/mss.c (hunk 2 above)
Comment (tsvisabo) [reply mam/mss/mss.c]: indeed

File: mam/api/api.c (hunk 5 above)
Comment (tsvisabo) [reply mam/api/api.c]: it could, and it's indeed better

File: mam/mam/channel.h [hunk 6]
```diff
@@ -76,6 +76,14 @@ retcode_t mam_channel_create(mam_prng_t const *const prng,
                              trits_t const channel_name,
                              mam_channel_t *const channel);
 
+/**
+ * Returns the number of remaining secret keys (unused leaves on merkle tree)
+ * @param channel The channel
+ *
+ * @return a status code
```
Comment (thibault-martinez) [inline mam/mam/channel.h]: Not a status code

File: mam/mam/endpoint.h [hunk 7]
```diff
@@ -73,6 +73,14 @@ retcode_t mam_endpoint_create(mam_prng_t const *const prng,
                               trits_t const endpoint_name,
                               mam_endpoint_t *const endpoint);
 
+/**
+ * Returns the number of remaining secret keys (unused leaves on merkle tree)
+ * @param endpoint The endpoint
+ *
+ * @return a status code
```
Comment (thibault-martinez) [inline mam/mam/endpoint.h]: Not a status code

File: mam/api/tests/test_api.c [hunk 8]
```diff
@@ -520,37 +521,44 @@ static void test_api_save_load() {
...
-    }
-  }
+  clean_channels();
   TEST_ASSERT(mam_api_destroy(&api) == RC_OK);
 
+  TEST_ASSERT(mam_api_init(&api, API_SEED) == RC_OK);
+  test_api_create_channels(&api, &cha, &ch1a, &epa, &ep1a, 4);
+  RUN_TEST(test_api_multiple_packets);
+  clean_channels();
+
```
Comment (thibault-martinez) [inline mam/api/tests/test_api.c:563]: You are missing a `mam_api_destroy `

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

**Review by thibault-martinez — APPROVED** [review]

**Pull request merged**